### PR TITLE
Fix #2756: Clickable text summary for Inactive Card

### LINKED ISSUE #2756: High fi StoryActivity Inactive Card
The mock for inactive card in `StoryActivity` has been updated to include clickable text to start the previous chapter.

https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/5abb0ace-19ee-48b8-8154-919154f924d2/

### PR DESCRIPTION
**Pull request by yashraj-01**
## Explanation
Fixes #2756. Made part of summary text clickable for Inactive Cards to scroll to the previous chapter.

**Screenshots of espresso tests passing**

`testStoryFragment_clickPrerequisiteChapter_prerequisiteChapterCardIsDisplayed()`
Screenshot from 2021-03-19 12-55-20

`testStoryFragment_configChange_clickPrerequisiteChapter_prerequisiteChapterCardIsDisplayed()`
Screenshot from 2021-03-19 12-56-58

`All tests`
Screenshot from 2021-03-19 12-57-12

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by yashraj-01** [comment]
@rt4914 PTAL.

**Review by anandwana001 — CHANGES_REQUESTED** [review]
@yashraj-iitr Add test case for testing this new behaviour in `StoryActivityTest`

Reference - 
https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing

File: app/src/main/java/org/oppia/android/app/story/StoryFragmentPresenter.kt [hunk 1]
```diff
@@ -142,6 +150,30 @@ class StoryFragmentPresenter @Inject constructor(
...
             )
+          if (storyItemViewModel.chapterSummary.chapterPlayState
+            == ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES
+          ) {
+            val missingPrerequisiteSummary = fragment.getString(
+              R.string.chapter_prerequisite_title_label,
+              storyItemViewModel.index.toString(),
+              storyItemViewModel.missingPrerequisiteChapter.name
+            )
+            val ss = SpannableString(missingPrerequisiteSummary)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/story/StoryFragmentPresenter.kt]: Rename ss to `chapterLockedSpannable` or something which can explain the context.

**Review by rt4914 — CHANGES_REQUESTED** [review]
@yashraj-iitr Thanks for the PR.

Three suggestions:
1. This is not working for `landscape` mode. So suggested making changes to other xml files too.
2. Add tests for the same.
3. The text is slightly cut-off on Pixel 3 device.

File: app/src/main/java/org/oppia/android/app/story/StoryFragmentPresenter.kt (hunk 1 above)
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/story/StoryFragmentPresenter.kt]: Done.

**Comment by yashraj-01** [comment]
> @yashraj-iitr Thanks for the PR.
> 
> Three suggestions:
> 
>     1. This is not working for `landscape` mode. So suggested making changes to other xml files too.
> 
>     2. Add tests for the same.
> 
>     3. The text is slightly cut-off on Pixel 3 device.
> 
> 
> 

I have applied the changes to other layout files. PTAL.

I will write the tests after I go through the reference. @rt4914 if you could provide me with some details about what the tests should check for, it will be really helpful.

Thanks.

**Comment by yashraj-01** [comment]
@rt4914 @anandwana001 I have written tests to test this new behaviour. PTAL. Thanks.

File: app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt [hunk 2]
```diff
@@ -485,6 +562,58 @@ class StoryFragmentTest {
...
+            val start = spannableString.getSpanStart(spanCandidate)
+            val end = spannableString.getSpanEnd(spanCandidate)
+            val sequence = spannableString.subSequence(start, end)
+            if (textToClick.toString().contains(sequence.toString())) {
+              span.onClick(textView)
+              return
+            }
+          }
+        }
+
```
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt]: remove empty line

File: app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt [hunk 3]
```diff
@@ -435,6 +445,73 @@ class StoryFragmentTest {
     }
   }
 
+  @Test
+  fun testStoryFragment_scrollToPrerequisiteChapter() {
```
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt]:
```kotlin
fun testStoryFragment_clickPrerequisiteChapter_prerequisiteChapterCardIsDisplayed()
```

File: app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt [hunk 4]
```diff
@@ -435,6 +445,73 @@ class StoryFragmentTest {
...
+      ).check(
+        matches(
+          isDisplayed()
+        )
+      )
+    }
+  }
+
+  @Test
+  fun testStoryFragment_changeConfiguration_scrollToPrerequisiteChapter() {
```
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt]:
```kotlin
fun testStoryFragment_configChange_clickPrerequisiteChapter_prerequisiteChapterCardIsDisplayed()
```

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Thanks, @yashraj-iitr Tests looks good to me. 
Suggested nit changes 
Also, please add a screenshot in PR description showing these tests are passing on Espresso.

**Comment by yashraj-01** [comment]
> Thanks, @yashraj-iitr Tests looks good to me.
> Suggested nit changes
> Also, please add a screenshot in PR description showing these tests are passing on Espresso.

@anandwana001 I have made the changes and also added the screenshots in the PR description. PTAL.

Also, there is one failing test with `NullPointerException`
Screenshot from 2021-03-18 13-35-49

I think this is because the `chapter_thumbnail` is not visible in the `landscape` configuration, thus resulting in `NullPointerException`. Please share your thoughts on this.
Screenshot_20210318-135133_Oppia


Thanks.

**Comment by anandwana001** [comment]
We should remove this test `testStoryFragment_configChange_chapterMissingPrerequisiteThumbnailIsBlurred` because as per the mocks, we don't show thumbnails in the landscape mode, but the chapter clicking thing should work on both portrait and landscape.

**Comment by yashraj-01** [comment]
> We should remove this test `testStoryFragment_configChange_chapterMissingPrerequisiteThumbnailIsBlurred` because as per the mocks, we don't show thumbnails in the landscape mode, but the chapter clicking thing should work on both portrait and landscape.

Yes. @anandwana001 should I remove the test `testStoryFragment_configChange_chapterMissingPrerequisiteThumbnailIsBlurred` ?

Also, the chapter summary click is working in both landscape as well as portrait mode.
Thank you.

**Comment by anandwana001** [comment]
Yeah, you can remove it.

**Comment by yashraj-01** [comment]
> Yeah, you can remove it.

@anandwana001 I have removed `testStoryFragment_configChange_chapterMissingPrerequisiteThumbnailIsBlurred` . 
PTAL. Thanks.

File: app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt [hunk 5]
```diff
@@ -489,6 +536,55 @@ class StoryFragmentTest {
     }
   }
 
+  fun clickClickableSpan(textToClick: CharSequence): ViewAction {
```
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt]: let make this function private as this is useable only in this file. If in future multiple test files need this function we will shift to utility.

Comment (anandwana001) [reply app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt]: Also, could we add a reference link which can act like a guide to this function?
https://stackoverflow.com/questions/38314077/how-to-click-a-clickablespan-using-espresso

**Review by anandwana001 — CHANGES_REQUESTED** [review]
just a nit update before the merge. Also, please update the screenshots in description.

File: app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt (hunk 5 above)
Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt]: Sure. I'll do it.

**Comment by yashraj-01** [comment]
@anandwana001 I have made the nit changes and also updated the PR description to include screenshots of tests passing. PTAL.
Thanks.

File: app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt (hunk 5 above)
Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt]: Done.

**Review by anandwana001 — APPROVED** [review]
LGTM, Thanks @yashraj-iitr

**Comment by yashraj-01** [comment]
Thanks, @anandwana001 for reviewing.
@rt4914 PTAL.

File: app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt (hunk 5 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt]: Resolving this as it is a small change.

**Comment by rt4914** [comment]
@yashraj-iitr I will have to close this PR as there is a **force-push**. Ideally we should not do a force push as it tampers the commit history and makes it hard to revisit the code. [Reference](https://github.com/oppia/oppia-android/wiki#address-review-comments-until-all-reviewers-give-lgtm-looks-good-to-me)
You can create a new PR for this. Also the 3rd. point in this comment has not been addressed or solved: https://github.com/oppia/oppia-android/pull/2885#pullrequestreview-612486378

**Comment by yashraj-01** [comment]
> @yashraj-iitr I will have to close this PR as there is a **force-push**. Ideally we should not do a force push as it tampers the commit history and makes it hard to revisit the code. [Reference](https://github.com/oppia/oppia-android/wiki#address-review-comments-until-all-reviewers-give-lgtm-looks-good-to-me)
> You can create a new PR for this. Also the 3rd. point in this comment has not been addressed or solved: [#2885 (review)](https://github.com/oppia/oppia-android/pull/2885#pullrequestreview-612486378)

@rt4914 should I create a new branch and make the same changes or is there another procedure? Please guide me.
Thanks.

**Comment by rt4914** [comment]
> > @yashraj-iitr I will have to close this PR as there is a **force-push**. Ideally we should not do a force push as it tampers the commit history and makes it hard to revisit the code. [Reference](https://github.com/oppia/oppia-android/wiki#address-review-comments-until-all-reviewers-give-lgtm-looks-good-to-me)
> > You can create a new PR for this. Also the 3rd. point in this comment has not been addressed or solved: [#2885 (review)](https://github.com/oppia/oppia-android/pull/2885#pullrequestreview-612486378)
> 
> @rt4914 should I create a new branch and make the same changes or is there another procedure? Please guide me.
> Thanks.

@yashraj-iitr You can create a new branch with same changes.

**Pull request closed without merging**